linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -4208,27 +4222,65 @@ private[optimizer] abstract class OptimizerCore(config: CommonPhaseConfig) {
     } else if (mutable) {
       withDedicatedVar(RefinedType(declaredType))
     } else {
-      val refinedType = value.tpe
+      def computeRefinedType(): RefinedType = bindingName match {
```

Comment:
OK, I think I understand now the constraints that lead to this solution :) Thoughts:

- From the code alone it is non-obvious that a call to `isSubtype` must be avoided if possible due to incremental tracking. Add a comment?
- It might be worth investigating two different subtypes of `Binding` for local / this. Notably, this bindings:
  - are always immutable
  - don't really have `declaredType`s, but `receiverType` (the way I understand `declaredType` is the type of the `This()` tree at definition site, so an uninteresting type as far as the optimizer is concerned; or is it interesting for the Long coercions and actually broken? :P).
  Although its unclear to me if this would make the code simpler at the end.

Reply:
> From the code alone it is non-obvious that a call to `isSubtype` must be avoided if possible due to incremental tracking. Add a comment?

I've significantly expanded in the comments. Hopefully this is clearer now.

> don't really have `declaredType`s, but `receiverType` (the way I understand `declaredType` is the type of the `This()` tree at definition site, so an uninteresting type as far as the optimizer is concerned; or is it interesting for the Long coercions and actually broken? :P).

Hum, I disagree. To me this is truly the declared type of the `this` "parameter" of the method. When we have
```scala
class Foo {
  def bar(x: Foobar): int
}
```
the parameter `x` has a declared type of `Foobar`, and the "parameter" `this` has type `Foo`. As if it were declared as
```scala
def bar(this: Foo, x: Foobar): int
```
(in a sense, this is something that Python makes quite explicit, although it doesn't have static types, with the `self` parameter of instance methods)